File: src/main/java/io/wdsj/asw/listener/BookListener.java
package io.wdsj.asw.listener;

import io.wdsj.asw.AdvancedSensitiveWords;
import io.wdsj.asw.event.ASWFilterEvent;
import io.wdsj.asw.event.EventType;
import io.wdsj.asw.impl.list.AdvancedList;
import io.wdsj.asw.manage.notice.Notifier;
import io.wdsj.asw.manage.punish.Punishment;
import io.wdsj.asw.setting.PluginMessages;
import io.wdsj.asw.setting.PluginSettings;
import io.wdsj.asw.util.Utils;
import io.wdsj.asw.util.cache.BookCache;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerEditBookEvent;
import org.bukkit.inventory.meta.BookMeta;

import java.util.List;

import static io.wdsj.asw.AdvancedSensitiveWords.*;
import static io.wdsj.asw.util.TimingUtils.addProcessStatistic;
import static io.wdsj.asw.util.Utils.messagesFilteredNum;

public class BookListener implements Listener {
    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onBook(PlayerEditBookEvent event) {
        if (!isInitialized) return;
        if (!settingsManager.getProperty(PluginSettings.ENABLE_BOOK_EDIT_CHECK)) return;
        Player player = event.getPlayer();
        if (player.hasPermission("advancedsensitivewords.bypass")) return;
        boolean isCacheEnabled = settingsManager.getProperty(PluginSettings.BOOK_CACHE);
        boolean skipReturnLine = settingsManager.getProperty(PluginSettings.BOOK_IGNORE_NEWLINE);
        String outMessage = "";
        String processedOutMessage = "";
        List<String> outList = new AdvancedList<>();
        List<String> originalPages = event.getNewBookMeta().getPages();
        boolean shouldSendMessage = false;
        BookMeta bookMeta = event.getNewBookMeta();
        int pageIndex = 1;
        long startTime = System.currentTimeMillis();
        if (bookMeta.hasPages()) {
            for (String originalPage : originalPages) {
                if (skipReturnLine) originalPage = originalPage.replace("\n", "").replace("§0", "");
                if (settingsManager.getProperty(PluginSettings.IGNORE_FORMAT_CODE)) originalPage = originalPage.replaceAll(Utils.getIgnoreFormatCodeRegex(), "");
                boolean isBookCached = BookCache.isBookCached(originalPage);
                List<String> censoredWordList = isBookCached && isCacheEnabled ? BookCache.getCachedBookSensitiveWordList(originalPage) : AdvancedSensitiveWords.sensitiveWordBs.findAll(originalPage);
                if (!censoredWordList.isEmpty()) {
                    String processedPage = isBookCached && isCacheEnabled ? BookCache.getCachedProcessedBookContent(originalPage) : AdvancedSensitiveWords.sensitiveWordBs.replace(originalPage);
                    if (!isBookCached && isCacheEnabled) BookCache.addToBookCache(originalPage, processedPage, censoredWordList);
                    if (settingsManager.getProperty(PluginSettings.BOOK_METHOD).equalsIgnoreCase("cancel")) {
                        event.setCancelled(true);
                        shouldSendMessage = true;
                        outMessage = originalPage;
                        outList = censoredWordList;
                        processedOutMessage = processedPage;
                        break;
                    }

                    bookMeta.setPage(pageIndex++, processedPage);
                    shouldSendMessage = true;
                    outMessage = originalPage;
                    outList = censoredWordList;
                    processedOutMessage = processedPage;
                }
            }
        }
        String originalAuthor = event.getNewBookMeta().getAuthor();
        if (originalAuthor != null) {
            if (settingsManager.getProperty(PluginSettings.IGNORE_FORMAT_CODE)) originalAuthor = originalAuthor.replaceAll(Utils.getIgnoreFormatCodeRegex(), "");
            List<String> censoredWordListAuthor = AdvancedSensitiveWords.sensitiveWordBs.findAll(originalAuthor);
            if (!censoredWordListAuthor.isEmpty()) {
                String processedAuthor = AdvancedSensitiveWords.sensitiveWordBs.replace(originalAuthor);
                if (settingsManager.getProperty(PluginSettings.BOOK_METHOD).equalsIgnoreCase("cancel")) {
                    event.setCancelled(true);
                } else {
                    bookMeta.setAuthor(processedAuthor);
                }
                shouldSendMessage = true;
                outMessage = originalAuthor;
                outList = censoredWordListAuthor;
                processedOutMessage = processedAuthor;
            }
        }


        String originalTitle = event.getNewBookMeta().getTitle();
        if (originalTitle != null) {
            if (settingsManager.getProperty(PluginSettings.IGNORE_FORMAT_CODE)) originalTitle = originalTitle.replaceAll(Utils.getIgnoreFormatCodeRegex(), "");
            List<String> censoredWordListTitle = AdvancedSensitiveWords.sensitiveWordBs.findAll(originalTitle);
            if (!censoredWordListTitle.isEmpty()) {
                String processedTitle = AdvancedSensitiveWords.sensitiveWordBs.replace(originalTitle);
                if (settingsManager.getProperty(PluginSettings.BOOK_METHOD).equalsIgnoreCase("cancel")) {
                    event.setCancelled(true);
                } else {
                    bookMeta.setTitle(processedTitle);
                }
                shouldSendMessage = true;
                outMessage = originalTitle;
                outList = censoredWordListTitle;
                processedOutMessage = processedTitle;
            }
        }

        if (shouldSendMessage) {
            event.setNewBookMeta(bookMeta);
            messagesFilteredNum.getAndIncrement();
            if (settingsManager.getProperty(PluginSettings.ENABLE_API)) {
                Bukkit.getPluginManager().callEvent(new ASWFilterEvent(player, outMessage, processedOutMessage, outList, EventType.BOOK, false));
            }
            long endTime = System.currentTimeMillis();
            addProcessStatistic(endTime, startTime);
            if (settingsManager.getProperty(PluginSettings.NOTICE_OPERATOR)) Notifier.notice(player, EventType.BOOK, outMessage);
            Punishment.punish(player);
        }

        if (settingsManager.getProperty(PluginSettings.BOOK_SEND_MESSAGE) && shouldSendMessage) {
            player.sendMessage(ChatColor.translateAlternateColorCodes('&', messagesManager.getProperty(PluginMessages.MESSAGE_ON_BOOK)));
        }

        if (settingsManager.getProperty(PluginSettings.LOG_VIOLATION) && shouldSendMessage) {
            Utils.logViolation(player.getName() + "(IP: " + Utils.getPlayerIp(player) + ")(Book)", outMessage + outList);
        }
    }
}

File: bukkit/src/main/kotlin/io/wdsj/asw/bukkit/listener/BookListener.kt
package io.wdsj.asw.bukkit.listener

import io.wdsj.asw.bukkit.AdvancedSensitiveWords.sensitiveWordBs
import io.wdsj.asw.bukkit.AdvancedSensitiveWords.settingsManager
import io.wdsj.asw.bukkit.setting.PluginMessages
import io.wdsj.asw.bukkit.setting.PluginSettings
import io.wdsj.asw.bukkit.type.ModuleType
import io.wdsj.asw.bukkit.util.PlayerProcessingGuard
import io.wdsj.asw.bukkit.util.Utils
import io.wdsj.asw.bukkit.util.ViolationReporter
import io.wdsj.asw.bukkit.util.cache.BookCache
import io.wdsj.asw.bukkit.util.message.MessageUtils
import org.bukkit.inventory.meta.BookMeta
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerEditBookEvent

class BookListener : Listener {
    private val processingGuard = PlayerProcessingGuard()
    private val violationReporter = ViolationReporter()

    @EventHandler(priority = EventPriority.LOWEST)
    fun onBook(event: PlayerEditBookEvent) {
        if (!settingsManager.getProperty(PluginSettings.ENABLE_BOOK_EDIT_CHECK)) return

        val player = event.player
        if (processingGuard.shouldSkipBasic(player)) return

        val startTime = System.currentTimeMillis()
        val bookMeta = event.newBookMeta
        var violation = censorPages(event, bookMeta)

        if (violation == null) {
            violation = censorCrossPage(event)
        }
        censorAuthor(event, bookMeta)?.let { violation = it }
        censorTitle(event, bookMeta)?.let { violation = it }
        val finalViolation = violation ?: return

        event.newBookMeta = bookMeta

        if (settingsManager.getProperty(PluginSettings.BOOK_SEND_MESSAGE)) {
            MessageUtils.sendMessage(player, PluginMessages.MESSAGE_ON_BOOK)
        }

        violationReporter.report(
            player = player,
            moduleType = ModuleType.BOOK,
            content = finalViolation.content,
            censoredWords = finalViolation.censoredWords,
            logSource = "Book",
            startTime = startTime,
            punish = settingsManager.getProperty(PluginSettings.BOOK_PUNISH),
        )
    }

    private fun censorPages(event: PlayerEditBookEvent, bookMeta: BookMeta): BookViolation? {
        if (!bookMeta.hasPages()) return null

        var violation: BookViolation? = null
        for ((pageIndex, page) in event.newBookMeta.pages.withIndex()) {
            val originalPage = preprocessPage(page)
            val censoredWords = findPageCensoredWords(originalPage)
            if (censoredWords.isEmpty()) continue

            violation = BookViolation(originalPage, censoredWords)
            if (isCancelMode()) {
                event.isCancelled = true
                break
            }

            bookMeta.setPage(pageIndex + 1, replacePage(originalPage, censoredWords))
        }
        return violation
    }

    private fun censorCrossPage(event: PlayerEditBookEvent): BookViolation? {
        if (!settingsManager.getProperty(PluginSettings.BOOK_CROSS_PAGE)) return null

        val originalPageCrossed = preprocessPage(event.newBookMeta.pages.joinToString(""))
        val censoredWords = sensitiveWordBs.findAll(originalPageCrossed)
        if (censoredWords.isEmpty()) return null

        event.isCancelled = true
        return BookViolation(originalPageCrossed, censoredWords)
    }

    private fun censorAuthor(event: PlayerEditBookEvent, bookMeta: BookMeta): BookViolation? {
        val author = event.newBookMeta.author ?: return null
        val originalAuthor = preprocessText(author)
        val censoredWords = sensitiveWordBs.findAll(originalAuthor)
        if (censoredWords.isEmpty()) return null

        if (isCancelMode()) {
            event.isCancelled = true
        } else {
            bookMeta.author = sensitiveWordBs.replace(originalAuthor)
        }
        return BookViolation(originalAuthor, censoredWords)
    }

    private fun censorTitle(event: PlayerEditBookEvent, bookMeta: BookMeta): BookViolation? {
        val title = event.newBookMeta.title ?: return null
        val originalTitle = preprocessText(title)
        val censoredWords = sensitiveWordBs.findAll(originalTitle)
        if (censoredWords.isEmpty()) return null

        if (isCancelMode()) {
            event.isCancelled = true
        } else {
            bookMeta.title = sensitiveWordBs.replace(originalTitle)
        }
        return BookViolation(originalTitle, censoredWords)
    }

    private fun findPageCensoredWords(page: String): List<String> {
        if (!settingsManager.getProperty(PluginSettings.BOOK_CACHE)) {
            return sensitiveWordBs.findAll(page)
        }
        if (BookCache.isBookCached(page)) {
            return BookCache.getCachedBookSensitiveWordList(page)
        }
        return sensitiveWordBs.findAll(page)
    }

    private fun replacePage(page: String, censoredWords: List<String>): String {
        if (!settingsManager.getProperty(PluginSettings.BOOK_CACHE)) {
            return sensitiveWordBs.replace(page)
        }
        if (BookCache.isBookCached(page)) {
            return BookCache.getCachedProcessedBookContent(page)
        }

        val processedPage = sensitiveWordBs.replace(page)
        BookCache.addToBookCache(page, processedPage, censoredWords)
        return processedPage
    }

    private fun preprocessPage(text: String): String {
        val lineNormalized = if (settingsManager.getProperty(PluginSettings.BOOK_IGNORE_NEWLINE)) {
            text.replace("\n", "").replace("§0", "")
        } else {
            text
        }
        return preprocessText(lineNormalized)
    }

    private fun preprocessText(text: String): String {
        if (!settingsManager.getProperty(PluginSettings.PRE_PROCESS)) return text
        return text.replace(Utils.preProcessRegex.toRegex(), "")
    }

    private fun isCancelMode(): Boolean {
        return settingsManager.getProperty(PluginSettings.BOOK_METHOD).equals("cancel", ignoreCase = true)
    }

    private data class BookViolation(
        val content: String,
        val censoredWords: List<String>,
    )
}
